docker/experimental/Dockerfile

```diff
@@ -0,0 +1,122 @@
+ARG python_version=3.8
+ARG debian_version=buster
+
+ARG cuda=0
+ARG cuda_version=11.2
+ARG cudnn_version=8.1.1.33
+# Debian repo doesn't have libcudnn
+ARG cuda_repo=ubuntu1804
+ARG tf_cuda_compute_capabilities="7.0"
+
+ARG tpuvm=1
+ARG build_cpp_tests=0
+
+# Contains all build dependencies for PT/XLA
+FROM python:${python_version}-${debian_version} AS builder
+
+RUN apt-get update
+RUN apt-get install -y curl gnupg libomp5 libopenblas-dev
+RUN pip install numpy pyyaml
+
+ARG cuda
+ARG cuda_version
+ARG cudnn_version
+ARG cuda_repo
+COPY docker/experimental/maybe_install_cuda.sh .
+RUN CUDA=${cuda} CUDA_REPO=${cuda_repo} CUDA_VERSION=${cuda_version} CUDNN_VERSION=${cudnn_version} ./maybe_install_cuda.sh build
+RUN rm maybe_install_cuda.sh
+ENV LD_LIBRARY_PATH=${cuda:+"${LD_LIBRARY_PATH}:/usr/local/cuda-${cuda_version}/targets/x86_64-linux/lib:/usr/local/nvidia/lib64"}
+
+ARG debian_version
+RUN echo "deb http://apt.llvm.org/${debian_version}/ llvm-toolchain-${debian_version}-8 main" >> /etc/apt/sources.list
+RUN echo "deb-src http://apt.llvm.org/${debian_version}/ llvm-toolchain-${debian_version}-8 main" >> /etc/apt/sources.list
+RUN curl https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+
+RUN apt-get update
+RUN apt-get -y install clang-8 clang++-8
+RUN update-alternatives --install /usr/bin/clang++ clang++ $(which clang++-8) 70
+
+ENV CC=clang-8
+ENV CXX=clang++-8
+
+RUN pip install mkl mkl-include setuptools typing_extensions cmake
+
+RUN git clone --recursive --depth=1 https://github.com/pytorch/pytorch.git
+WORKDIR /pytorch
+COPY torch_patches/ torch_patches/
+RUN bash -c "torch_pin=$(cat torch_patches/.torch_pin &); git fetch origin ${torch_pin:-master}; git checkout origin/${torch_pin:-master}"
+
+RUN USE_CUDA=0 python setup.py bdist_wheel
+
+RUN pip install dist/*.whl
+
+RUN curl -LO https://github.com/bazelbuild/bazelisk/releases/download/v1.11.0/bazelisk-linux-amd64
+RUN mv bazelisk-linux-amd64 /usr/local/bin/bazel
+RUN chmod +x /usr/local/bin/bazel
+
+RUN mkdir xla
```

Comment:
I a, a bit lost here, seems like we didn't git clone xla repo, so where would the code come from?

Reply:
You run this build from a working copy of PyTorch/XLA, so each of the files/directories needed for the build are copied in from the working directory in the `COPY` statements below. Each `COPY` in the `build` step only brings in what is needed to make the build incremental -- if any of the files in the `COPY` are different, all of the `RUN` steps have to run again and can't be cached. If you copied in everything at once, changes in `torch_xla` would force you to rebuild TensorFlow, for example.

The `builder` image doesn't have any PT/XLA checked out. My idea here is that you can check out PT/XLA on the host machine and mount it at `/pytorch/xla` in the container to have just the build dependencies preinstalled.

The whole working copy is copied into the `release` image at the end in `COPY . /pytorch/xla`.